### PR TITLE
Turns API

### PR DESCRIPTION
**Pull request by patshone-gsl**

### CODE REVIEW

File: app/Http/Controllers/API/IntentsController.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        $this->middleware('auth');
+    }
+
+    /**
+     * Display the specified Intent.
+     *
+     * @param Intent $Intent
+     * @reIntent IntentResource
+     */
+    public function show(Intent $Intent): IntentResource
```
Comment (stu-greenshoots) [inline app/Http/Controllers/API/IntentsController.php:36]: Why do we capitalize the `I` in intent variables?

File: app/Http/Controllers/API/TurnsController.php [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        $responseIntents = ConversationDataClient::getAllResponseIntentsByTurn($turn, false);
+
+        $turnIntents = [];
+        foreach ($requestIntents as $intent) {
+            $turnIntents[] = new TurnIntentResource($intent, 'REQUEST');
+        }
+        foreach ($responseIntents as $intent) {
+            $turnIntents[] = new TurnIntentResource($intent, 'RESPONSE');
+        }
+        return response()->json(new TurnIntentResourceCollection($turnIntents));
```
Comment (stu-greenshoots) [inline app/Http/Controllers/API/TurnsController.php:52]: Can we just return the ResourceCollection here?

File: app/Http/Controllers/API/TurnsController.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    public function destroy(Turn $Turn): Response
+    {
+        if (ConversationDataClient::deleteTurnByUid($Turn->getUid())) {
+            return response()->noContent(200);
+        } else {
+            return response('Error deleting conversation, check the logs', 500);
+        }
+    }
+
+    public function getTurnIntentByTurnAndIntent(Turn $turn, Intent $intent) : TurnIntentResource
```
Comment (stu-greenshoots) [inline app/Http/Controllers/API/TurnsController.php:117]: phpdoc :)

File: app/Http/Controllers/API/TurnsController.php [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        $turnWithIntent = ConversationDataClient::getTurnWithIntent($turn->getUid(), $intent->getUid());
+        if ($turnWithIntent->getRequestIntents()->count() > 0) {
+            return new TurnIntentResource($turnWithIntent->getRequestIntents()->first(), 'REQUEST');
```
Comment (stu-greenshoots) [inline app/Http/Controllers/API/TurnsController.php:121]: Is the request / response type not stored against the intent? Why do we have to work it out like this?

File: app/Http/Controllers/API/TurnsController.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        } elseif ($turnWithIntent->getRequestIntents()->count() > 0) {
```
Comment (stu-greenshoots) [inline app/Http/Controllers/API/TurnsController.php:122]: can this just be an else, or outside of the if block completely?

File: app/Http/Controllers/API/TurnsController.php [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    public function updateTurnIntent(TurnIntentRequest $request, Turn $turn, Intent $intent) :
+    TurnIntentResource
+    {
+        $content = $request->json()->all();
+        $order = $content['order'];
+        $patchIntent = Serializer::denormalize($content['intent'], Intent::class, 'json');
+        $patchIntent->setUid($intent->getUid());
+        // First update the intent data
+        $updatedIntent = ConversationDataClient::updateIntent($patchIntent);
```
Comment (stu-greenshoots) [inline app/Http/Controllers/API/TurnsController.php:135]: Is there a way to do this transactionally? What happens if something goes wrong in the middle?

File: app/Http/Controllers/API/TurnsController.php [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        } elseif ($turnWithIntent->getRequestIntents()->count() > 0) {
+            return new TurnIntentResource($turnWithIntent->getResponseIntents()->first(), 'RESPONSE');
+        }
+    }
+
+    public function updateTurnIntent(TurnIntentRequest $request, Turn $turn, Intent $intent) :
+    TurnIntentResource
+    {
+        $content = $request->json()->all();
+        $order = $content['order'];
```
Comment (stu-greenshoots) [inline app/Http/Controllers/API/TurnsController.php:131]: This isn't used (accessed again in the updateTurnIntentRelation call.

Also, we can just use `$request->order` directly if we want

File: app/Http/Resources/FocusedIntentResource.php [hunk 8]
```diff
@@ -41,6 +87,42 @@ public function toArray($request)
     {
         // reshape the response renaming conversation to focussedConversation
         $normalizedIntent = Serializer::normalize($this->resource, 'json', self::$fields);
-        return $normalizedIntent;
+
+        $normalizedIntentToReturn = [];
```
Comment (stu-greenshoots) [inline app/Http/Resources/FocusedIntentResource.php:91]: This hurts my head a bit :|

File: routes/api.php [hunk 9]
```diff
@@ -128,6 +128,33 @@
...
+            /** Todo */
+            Route::get('turns/{turn}/turn-intents/{intent}', 'TurnsController@getTurnIntentByTurnAndIntent');
+            Route::patch('turns/{turn}/turn-intents/{intent}', 'TurnsController@updateTurnIntent');
+            /** End Todo */
+
+            Route::get('intents/{intent}', 'IntentsController@show');
+            Route::patch('intents/{intent}', 'IntentsController@update');
+            Route::delete('intents/{intent}', 'IntentsController@destroy');
+
+            /** Todo */
```
Comment (stu-greenshoots) [inline routes/api.php:151]: Are these TODOs still valid?

File: app/Http/Resources/FocusedIntentResource.php (hunk 8 above)
Comment (patshone-gsl) [reply app/Http/Resources/FocusedIntentResource.php:91]: Agreed. I think we can change these resources by updating the front end - it feels unnecessary to flip everything round when really we have all the data we need from the original GraphQL response

File: routes/api.php (hunk 9 above)
Comment (patshone-gsl) [reply routes/api.php:151]: Nope, have removed

File: app/Http/Controllers/API/IntentsController.php (hunk 1 above)
Comment (patshone-gsl) [reply app/Http/Controllers/API/IntentsController.php:36]: Resolved

File: app/Http/Controllers/API/TurnsController.php (hunk 7 above)
Comment (patshone-gsl) [reply app/Http/Controllers/API/TurnsController.php:131]: Updated to use the request object directly

**Pull request merged**